libs/ledger-live-common/src/__tests__/hw/getDeviceInfo.ts

```diff
@@ -36,7 +36,7 @@ test("1.3.1", async () => {
   expect(res).toMatchObject({
     version: "1.3.1",
     mcuVersion: "1.1",
-    majMin: "1.3",
```

Comment:
we have made these tests in the past to make sure we don't accidentally break old device logic.
and the changes revealed by this test indicate possible breaking for past versions.

if you search for it, you will notice in some part of the firmware update process, we do

```
(deviceInfo.majMin === "0.6" || deviceInfo.majMin === "0.7")
```

so we have a strong expectation in our codebase that these fields are in form of `x.y` . ugly legacy :'( but i think we don't want to break firmware update for old device too much.

Reply:
If the old firmware version is in format of `x.y` w/o `.z`, I think my fix will always work, e.g. there are some test cases for this and I don't change them. If not, can you please give out a criteria with which I can identify it's an old version firmware ?

Reply:
it would be preferable to introduce a new field than rewrite this one tbh.
the name of the field is literally "majMin" as in "maj.min.patch" version of the full value (that full value is in version)

Reply:
I think this is the origin of a bug when updating firmware. Actually, we have changed the version from `x.y` to `x.y.z`. After installing new firmware, some mechanism will check whether the version downloaded (`x.y.z`) is the same to the version installed (was `x.y` before, returned by `getDeviceInfo()` in field `majMin`). The field `majMin` is an extrait of the complete version, if the old version is `1.2 - dev` for example, the field will be `1.2`, if the new version is `1.2.3 - rc2`, the field will be `1.2.3`. This will not impact the old version.

Reply:
@gre Actually the `version` is not always equal to `majMin` in my proposal. The `version` may be `x.y.z - dev` for example.

Reply:
i understand there is a bug on this that we need to fix for Stax 👍

However, i think the current changes are breaking: if you look in `libs/ledger-live-common/src/hw/firmwareUpdate-repair.ts` you will notice we have hardcoded logic that expects this field to be a `x.y`, that part is needed for old Ledger Nano S firmware to "repair" their firmware update. This is an old mechanism but as we are still manage the Nano S devices, **if** we want to change this part, we will have to be sure to be able to test it and the scope is wider than "fixing stax mcu update" i think. Typically we will need a bunch of old devices on old firmwares (and these are harder to find these days).


![Capture d’écran 2024-03-18 à 10 22 15](https://github.com/LedgerHQ/ledger-live/assets/211411/8f1c92f7-109b-4527-8abd-555b1131cfc7)

another location where this will break is very near the code you have modified:

```
const blVersionAliases = {
  "0.0": "0.6",
};
...
of(blVersionAliases[deviceInfo.majMin])
```

won't work anymore. this code is again a very old logic for very old firmware (if i recall, the very first firm of nano S had a specific case on this, i think some doc exists on confluence)

---

So i'm not exactly sure what is the best solution forward from here, but we could make the change less impacting by introducing a new field "majMinPatch" (can't figure out a better name, i was trying to figure out what is the semantic name in https://semver.org/ but they don't have specific names for the `x.y.z` part)

and then use this new field when it makes sense but then still have a "fallback" to the old logic when it doesn't find the data related to majMinPatch.

If we however, consider reworking this completely, we will have to test on many past firmwares on nominal and error flows (and that's challenging)
example of situation (2020) where this broke before https://ledgerhq.atlassian.net/wiki/spaces/HUB/pages/1269989380/Postmortem+LLD+1.8.0+broke+Firmware+Update+for+LNS+1.3.1